crates/derive/src/metrics/mod.rs

```diff
@@ -0,0 +1,34 @@
+//! Metrics for the derivation pipeline.
+
+mod noop;
+
+use alloc::sync::Arc;
+use core::fmt::Debug;
+
+use crate::traits::{DerivationPipelineMetrics, StepResult};
+
+/// Composite metrics struct containing metrics for all stages.
+pub struct PipelineMetrics {
+    pub(crate) derivation_pipeline_metrics: Arc<dyn DerivationPipelineMetrics + Send + Sync>,
+    // todo: add more metrics here for each stage
```

Comment:
Preference to create a sub-ticket in github issues over leaving TODO comments in the codebase.